arrows-stdlib-cats/src/main/scala/arrows/stdlib/cats.scala

```diff
@@ -0,0 +1,114 @@
+package arrows.stdlib
+
+import cats.{ Applicative, StackSafeMonad }
+import cats.effect._
+import cats.arrow.ArrowChoice
+import cats.kernel.{ Monoid, Semigroup }
+import cats.mtl.ApplicativeAsk
+
+import scala.concurrent.{ ExecutionContext, Promise }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success }
+
+object Cats extends ArrowInstances
+
+sealed abstract class ArrowInstances extends ArrowInstances1 {
+
+  implicit def catsEffectForTask(implicit ec: ExecutionContext): Effect[Task] = new ArrowAsync[Unit] with Effect[Task] {
+    def runAsync[A](fa: Task[A])(cb: Either[Throwable, A] => IO[Unit]): SyncIO[Unit] =
+      SyncIO { fa.run(())(ec); () }
+  }
+
+  implicit val catsArrowChoiceForArrow: ArrowChoice[Arrow] = new ArrowChoice[Arrow] {
+    def choose[A, B, C, D](f: Arrow[A, C])(g: Arrow[B, D]): Arrow[Either[A, B], Either[C, D]] =
+      Arrow[Either[A, B]].flatMap {
+        case Left(a)  => f(a).map(Left(_))
+        case Right(b) => g(b).map(Right(_))
+      }
+
+    def lift[A, B](f: A => B): Arrow[A, B] = Arrow[A].map(f)
+
+    def first[A, B, C](fa: Arrow[A, B]): Arrow[(A, C), (B, C)] =
+      Arrow[(A, C)].flatMap { case (a, c) => fa(a).map(_ -> c) }
+
+    def compose[A, B, C](f: Arrow[B, C], g: Arrow[A, B]): Arrow[A, C] = g.andThen(f)
+  }
+
+  implicit def catsApplicativeAskForArrow[E]: ApplicativeAsk[Arrow[E, ?], E] = new ApplicativeAsk[Arrow[E, ?], E] {
+    val applicative: Applicative[Arrow[E, ?]] = catsAsyncForArrow[E]
+
+    def ask: Arrow[E, E] = Arrow[E]
+
+    def reader[A](f: E => A): Arrow[E, A] = ask.map(f)
+
+  }
+
+  implicit def catsMonoidForArrow[A, B](implicit B0: Monoid[B]): Monoid[Arrow[A, B]] =
+    new Monoid[Arrow[A, B]] with ArrowSemigroup[A, B] {
+      implicit def B: Semigroup[B] = B0
+
+      def empty: Arrow[A, B] = Arrow.successful(Monoid[B].empty)
+    }
+}
+
+sealed abstract class ArrowInstances1 {
+
+  implicit def catsAsyncForArrow[E]: Async[Arrow[E, ?]] = new ArrowAsync[E] {}
+
+  implicit def catsSemigroupForArrow[A, B](implicit B0: Semigroup[B]): Semigroup[Arrow[A, B]] = new ArrowSemigroup[A, B] {
+    implicit def B: Semigroup[B] = B0
+  }
+}
+
+trait ArrowAsync[E] extends StackSafeMonad[Arrow[E, ?]] with Async[Arrow[E, ?]] {
+  def flatMap[A, B](fa: Arrow[E, A])(f: A => Arrow[E, B]): Arrow[E, B] =
+    Arrow[E].flatMap(e => fa(e).flatMap(a => f(a)(e)))
```

Comment:
Sorry, I'm still not sure this is a good approach. There's the performance issue, and I don't see why passing the same input to both arrows is useful other than when the arrow is a `Task` so the input is fixed. Could you give an example of what you couldn't express if this class used `Task` instead of `Arrow`?

Reply:
Sure, together with something like `ApplicativeAsk`, you can define cool `ReaderT` style code with `Arrow`:

```scala

def foo[F[_]: Monad: ApplicativeAsk[?[_], Config]] = for {
  config <- ApplicativeAsk[F, Config].ask
  response <- serviceCall(config)
} yield response
```

Just a small example, but this is a pretty neat feature IMO and one could easily use this with e.g. `http4s` which only needs `Sync` to define the routes :)